### ISSUE TITLE
[feature request] "frecency", nonrestrictive Autocomplete

### ISSUE THREAD

**Issue by sw-yx** [enquirer/enquirer#158]
this is doable in userland but i think it'd be a great built in thing.

basically i want to be able to let the user type, but autocomplete based on [frecency](https://github.com/mixmaxhq/frecency) of what they have typed before (like using a small cache defined by the developer), but also not be restricted by the list inside the autocomplete i.e. it behaves like an Input if the user wants.

does this make sense?

**Comment by doowb** [comment]
@sw-yx I like this idea a lot!

I think there are 2 parts that you're describing:

1. allowing a user to select their literal input if it doesn't exactly match an item in the provided choices.
    I don't remember off the top of my head, but this might already be available.

1. frecency: saving and scoring user's previous search and selection to sort results based on frequency and recency of the selections made.
    We've discussed saving answers (history) and this may be something that could be an option there, but I think some more discussion needs to happen. We're generally trying include features that allow implementers to implement these customizations (in userland as you mentioned) instead of embedding them in the prompts.

I think an example implementing this would be awesome, then we could discuss if and how it could be built in.

**Comment by sw-yx** [comment]
yeah for sure, just need to make time to build it 😅

**Comment by jonschlinkert** [comment]
I love both ideas, especially being able to sort suggestions by frecency. I'll focus on that feature here... 

In userland, I *think* this can be accomplished by just sorting the results that are returned by a custom `suggest` function. Something like:

```js
const { prompt } = require('enquirer');
const answers = await prompt({
  type: 'autocomplete',
  name: 'flavor',
  message: 'Pick your favorite flavor',
  suggest(input, choices) {
    // filter choices by user "input", and sort by "frecency"
  },
  choices: [
    'almond',
    'apple',
    'banana',
    'cherry',
    'chocolate',
    'cinnamon',
    'coconut',
  ]
});
```

If we add sort-by-frecency as a built-in feature, I think we can do something similar to the history feature in the [Input prompt](https://github.com/enquirer/enquirer/blob/master/lib/prompts/input.js), but we'd also store information makes sorting by frecency possible.

**Comment by sw-yx** [comment]
the reason i cant use Autocomplete for this is because it is restrictive, exactly as doowb describes, i need to be able to allow for input of new items not on the list.

but yea i just gotta find time to build it. maybe someone will beat me to it :)

**Comment by sw-yx** [comment]
more notes on frecency in Slack: https://slack.engineering/a-faster-smarter-quick-switcher-77cbc193cb60

and also rupa/z

**Comment by jonschlinkert** [comment]
> i need to be able to allow for input of new items not on the list.

K, got it - I think. The ability to add values on the fly AND have them be [sorted correctly](https://slack.engineering/a-faster-smarter-quick-switcher-77cbc193cb60#700c) in comparison to other values that are sorted by frecency is the key. Correct?

**Comment by sw-yx** [comment]
yeah. obviously adding a new value means we dont need to AutoComplete anymore and we just become an Input.  so it's kinda bimodal: if there's a match, be an Autocomplete, if there's no match, be an Input and save the answer in the history for future autocompleting

**Comment by sw-yx** [comment]
frecency isnt the best documented lib. i managed to get automcplete working with frecency: https://github.com/sw-yx/cli-state#accessglobalfrecency

but now i need to make it turn into an `Input` for unrecognized values.

**Comment by jonschlinkert** [comment]
> if there's no match, be an Input and save the answer in the history for future autocompleting

yeah, I like that, and it's kind of how I was thinking of it too. 

> but now i need to make it turn into an `Input` for unrecognized values.

We can help with that part, I think this feature might need to be built in. 

Just thinking out loud... FWIW the goal of the current release was to "get it working" so that usability was as close as what we wanted as possible. The goal of the next release is to make things like this easier for implementors to customize (but also to make the code easier to work with for maintainers and contributors). Ideally, we can find a way to allow prompts to use actions from other prompts so that it's possible to create "hybrid" prompts, like a choices-prompt that allows a choice to be added (which would have characteristics of an input prompt). Similar to what we want to accomplish with the autocomplete. A best-case scenario would be to create a single prompt that is composed of multiple other prompts, with any of the available types. The experience would be more like the forms we use in the real world, or on the web. 

The challenge (for me) is to figure out the best balance of abstraction and customizability. At the end of the day, I'm not sure we can get what we need without having more abstraction - for example, we probably need to decouple all actions (key combos and how they work for each type). Which makes the code harder to maintain and debug, but for the vast majority of users will result in a better experience. 

That's a lot of rambling, but I want to make sure we encourage contribution with code that's easy to understand, but also provide a good user experience, and any feedback would be welcome.

**Comment by jonschlinkert** [comment]
btw, @sw-yx thanks so much for the amazing and flattering tweet!

**Comment by sw-yx** [comment]
np! yeah i understand, take your time to get it right. this is a nice test usecase i suppose :)
